src/tiledb/cloud/geospatial/ingestion.py

```diff
@@ -0,0 +1,1405 @@
+    :param res: Tuple[float, float], output resolution in x/y
+    :param verbose: verbose logging, defaults to False
+    :param trace: bool, enabling log tracing, defaults to False
+    :param log_uri: log array URI
+    :return: A dict containing the kwargs needed for the next function call
+    """
+    from tiledb.cloud.utilities import Profiler
+    from tiledb.cloud.utilities import find
+    from tiledb.cloud.utilities import get_logger_wrapper
+    from tiledb.cloud.utilities import max_memory_usage
+
+    logger = get_logger_wrapper(verbose)
+    with Profiler(array_uri=log_uri, id=id, trace=trace):
+        try:
+
+            def raster_match(f: str):
+                import rasterio
+
+                try:
+                    rasterio.driver_from_extension(f)
+                    return True
+                except ValueError:
+                    return False
+
+            def pointcloud_match(f: str):
+                return os.path.splitext(f)[-1].lstrip(".").lower() in ["las", "laz"]
+
+            def geometry_match(f: str):
+                import fiona
+
+                try:
+                    fiona.drvsupport.driver_from_extension(f)
+                    return True
+                except ValueError:
+                    return False
+
+            fns = {
+                DatasetType.POINTCLOUD: {
+                    "pattern_fn": pointcloud_match,
+                    "meta_fn": load_pointcloud_metadata,
+                    "kwargs": {},
+                },
+                DatasetType.RASTER: {
+                    "pattern_fn": raster_match,
+                    "meta_fn": load_raster_metadata,
+                    "kwargs": {
+                        "pixels_per_fragment": pixels_per_fragment,
+                        "res": res,
+                    },
+                },
+                DatasetType.GEOMETRY: {
+                    "pattern_fn": geometry_match,
+                    "meta_fn": load_geometry_metadata,
+                    "kwargs": {},
+                },
+            }
+
+            kwargs = {}
+
+            if dataset_list_uri:
+                sources = read_uris(
+                    dataset_list_uri,
+                    dataset_type=dataset_type,
+                    log_uri=log_uri,
+                    config=config,
+                    max_files=max_files,
+                )
+
+            if search_uri:
+                sources = find(
+                    search_uri,
+                    config=config,
+                    exclude=ignore,
+                    include=pattern if pattern else fns[dataset_type]["pattern_fn"],
+                    max_count=max_files,
+                )
+
+            if not sources:
+                raise ValueError(f"No {dataset_type.name} datasets found")
+
+            meta_kwargs = fns[dataset_type]["kwargs"]
+            meta = fns[dataset_type]["meta_fn"](sources, config=config, **meta_kwargs)
```

Comment:
So in testing this on some larger scale use cases. Getting the metadata in-line is a problem. I believe we should batch the output into jobs of 10 or 100 files, and grab metadata in parallel in another stage of the task graph. I've seen it take over 30m when there is 6000 files.

Reply:
I think we should change this after this PR has been merged. There are a couple of scalability improvements we can make.

Reply:
Also just noting that geospatial datasets tend to be large as opposed to thousands of files. We should anticipate the problem but I would like to make this change as an incremental PR.

Reply:
Sure thing, thanks!